// @google/clasp/tests/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var os_1 = __importDefault(require("os"));
var path_1 = __importDefault(require("path"));
var functions_1 = require("./functions");
// Sample files
exports.TEST_CODE_JS = 'function test() { Logger.log(\'test\'); }';
exports.TEST_PAGE_HTML = '<html><body><p>hello there</p></body</html>';
exports.TEST_APPSSCRIPT_JSON_WITHOUT_RUN_API = JSON.stringify({
    timeZone: 'America/Los_Angeles',
    dependencies: {},
    exceptionLogging: 'STACKDRIVER',
});
exports.TEST_APPSSCRIPT_JSON_WITH_RUN_API = JSON.stringify({
    timeZone: 'America/Los_Angeles',
    dependencies: {},
    exceptionLogging: 'STACKDRIVER',
    executionApi: {
        access: 'MYSELF',
    },
});
// Travis Env Variables
exports.IS_PR = (process.env.TRAVIS_PULL_REQUEST === 'true');
exports.SCRIPT_ID = process.env.SCRIPT_ID || '';
exports.PROJECT_ID = process.env.PROJECT_ID || '';
var HOME = process.env.HOME || '';
// Paths
exports.CLASP_PATHS = {
    clientCredsLocal: 'client_credentials.json',
    rcGlobal: path_1.default.join(HOME, '.clasprc.json'),
    rcLocal: '.clasprc.json',
    settingsLocal: '.clasp.json',
};
// Other constants
exports.CLASP = (os_1.default.type() === 'Windows_NT') ? 'clasp.cmd' : 'clasp';
exports.CLASP_USAGE = 'Usage: clasp <command> [options]';
var VALID_CLASP_SETTINGS = {
    scriptId: exports.SCRIPT_ID,
    projectId: exports.PROJECT_ID,
};
var INVALID_CLASP_SETTINGS = {
    scriptId: exports.SCRIPT_ID,
    projectId: "project-id-" + functions_1.rndStr(),
};
var VALID_CLASP_SETTINGS_WITHOUT_PROJECT_ID = {
    scriptId: exports.SCRIPT_ID,
};
exports.CLASP_SETTINGS = {
    valid: JSON.stringify(VALID_CLASP_SETTINGS),
    invalid: JSON.stringify(INVALID_CLASP_SETTINGS),
    validWithoutProjectId: JSON.stringify(VALID_CLASP_SETTINGS_WITHOUT_PROJECT_ID),
};
var FAKE_CLASPRC_TOKEN = {
    access_token: functions_1.rndStr(),
    refresh_token: functions_1.rndStr(),
    scope: 'https://www.googleapis.com/auth/script.projects',
    token_type: 'Bearer',
    expiry_date: (new Date()).getTime(),
};
var oAuth2ClientOptions = {
    clientId: functions_1.rndStr() + ".apps.googleusercontent.com",
    clientSecret: functions_1.rndStr(),
};
var FAKE_CLASPRC_LOCAL = {
    token: FAKE_CLASPRC_TOKEN,
    oauth2ClientSettings: oAuth2ClientOptions,
    isLocalCreds: true,
};
exports.FAKE_CLASPRC = {
    local: JSON.stringify(FAKE_CLASPRC_LOCAL),
    token: JSON.stringify(FAKE_CLASPRC_TOKEN),
};
var FAKE_CLIENT_CREDS = {
    installed: {
        client_id: functions_1.rndStr() + ".apps.googleusercontent.com",
        client_secret: functions_1.rndStr(),
    },
};
var INVALID_CLIENT_CREDS = {
    installed: {
        client_id: functions_1.rndStr() + ".apps.googleusercontent.com",
    },
};
exports.CLIENT_CREDS = {
    fake: JSON.stringify(FAKE_CLIENT_CREDS),
    invalid: JSON.stringify(INVALID_CLIENT_CREDS),
};
