// @google/clasp/tests/functions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var constants_1 = require("./constants");
var fs_extra_1 = __importDefault(require("fs-extra"));
var path_1 = __importDefault(require("path"));
var tmp_1 = __importDefault(require("tmp"));
/** basic cleanup after tests */
exports.cleanup = function () {
    fs_extra_1.default.removeSync('.clasp.json');
    fs_extra_1.default.removeSync('.claspignore');
    fs_extra_1.default.removeSync('Code.js');
    fs_extra_1.default.removeSync('appsscript.json');
};
/** basic setup for tests */
exports.setup = function () {
    fs_extra_1.default.writeFileSync('.clasp.json', constants_1.CLASP_SETTINGS.valid);
    fs_extra_1.default.writeFileSync('appsscript.json', constants_1.TEST_APPSSCRIPT_JSON_WITHOUT_RUN_API);
};
/** setup for tests not using the run API */
exports.setupWithoutGCPProject = function () {
    fs_extra_1.default.writeFileSync('.clasp.json', constants_1.CLASP_SETTINGS.validWithoutProjectId);
    fs_extra_1.default.writeFileSync('appsscript.json', constants_1.TEST_APPSSCRIPT_JSON_WITHOUT_RUN_API);
};
/** setup for tests using the run API */
exports.setupWithRunManifest = function () {
    fs_extra_1.default.writeFileSync('.clasp.json', constants_1.CLASP_SETTINGS.valid);
    fs_extra_1.default.writeFileSync('appsscript.json', constants_1.TEST_APPSSCRIPT_JSON_WITH_RUN_API);
};
/** produce a pseudo random string */
exports.rndStr = function () { return Math.random().toString(36).substr(2); };
/**
 * backup clasp settings. Use `restoreSettings()` to restore these.
 */
exports.backupSettings = function () {
    if (fs_extra_1.default.existsSync(constants_1.CLASP_PATHS.rcGlobal)) {
        fs_extra_1.default.copyFileSync(constants_1.CLASP_PATHS.rcGlobal, constants_1.CLASP_PATHS.rcGlobal + "~");
    }
    if (fs_extra_1.default.existsSync(constants_1.CLASP_PATHS.rcLocal)) {
        fs_extra_1.default.copyFileSync(constants_1.CLASP_PATHS.rcLocal, constants_1.CLASP_PATHS.rcLocal + "~");
    }
    if (fs_extra_1.default.existsSync(constants_1.CLASP_PATHS.settingsLocal)) {
        fs_extra_1.default.copyFileSync(constants_1.CLASP_PATHS.settingsLocal, constants_1.CLASP_PATHS.settingsLocal + "~");
    }
};
/**
 * restore clasp settings backuped up using `backupSettings()`
 */
exports.restoreSettings = function () {
    if (fs_extra_1.default.existsSync(constants_1.CLASP_PATHS.rcGlobal + "~")) {
        fs_extra_1.default.renameSync(constants_1.CLASP_PATHS.rcGlobal + "~", constants_1.CLASP_PATHS.rcGlobal);
    }
    if (fs_extra_1.default.existsSync(constants_1.CLASP_PATHS.rcLocal + "~")) {
        fs_extra_1.default.renameSync(constants_1.CLASP_PATHS.rcLocal + "~", constants_1.CLASP_PATHS.rcLocal);
    }
    if (fs_extra_1.default.existsSync(constants_1.CLASP_PATHS.settingsLocal + "~")) {
        fs_extra_1.default.renameSync(constants_1.CLASP_PATHS.settingsLocal + "~", constants_1.CLASP_PATHS.settingsLocal);
    }
};
/**
 * create a temporary directory and its content, then return its path as a string
 *
 * @param {Array<{ file: string, data: string }} filepathsAndContents directory content (files)
 */
function setupTmpDirectory(filepathsAndContents) {
    fs_extra_1.default.ensureDirSync('tmp');
    var tmpdir = tmp_1.default.dirSync({ unsafeCleanup: true, dir: 'tmp/', keep: false }).name;
    filepathsAndContents.forEach(function (_a) {
        var file = _a.file, data = _a.data;
        fs_extra_1.default.outputFileSync(path_1.default.join(tmpdir, file), data);
    });
    return tmpdir;
}
exports.setupTmpDirectory = setupTmpDirectory;
