// @google/clasp/src/commands/deploy.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var auth_1 = require("../auth");
var utils_1 = require("../utils");
/**
 * Deploys an Apps Script project.
 * @param cmd.versionNumber {string} The project version to deploy at.
 * @param cmd.description   {string} The deployment description.
 * @param cmd.deploymentId  {string} The deployment ID to redeploy.
 */
exports.default = (function (cmd) { return __awaiter(void 0, void 0, void 0, function () {
    var scriptId, versionNumber, _a, description, deploymentId, version, deployments;
    return __generator(this, function (_b) {
        switch (_b.label) {
            case 0: return [4 /*yield*/, utils_1.checkIfOnline()];
            case 1:
                _b.sent();
                return [4 /*yield*/, auth_1.loadAPICredentials()];
            case 2:
                _b.sent();
                return [4 /*yield*/, utils_1.getProjectSettings()];
            case 3:
                scriptId = (_b.sent()).scriptId;
                if (!scriptId)
                    return [2 /*return*/];
                utils_1.spinner.setSpinnerTitle(utils_1.LOG.DEPLOYMENT_START(scriptId)).start();
                versionNumber = cmd.versionNumber;
                _a = cmd.description, description = _a === void 0 ? '' : _a, deploymentId = cmd.deploymentId;
                if (!!versionNumber) return [3 /*break*/, 5];
                return [4 /*yield*/, auth_1.script.projects.versions.create({
                        scriptId: scriptId,
                        requestBody: {
                            description: description,
                        },
                    })];
            case 4:
                version = _b.sent();
                utils_1.spinner.stop(true);
                if (version.status !== 200)
                    utils_1.logError(null, utils_1.ERROR.ONE_DEPLOYMENT_CREATE);
                versionNumber = version.data.versionNumber || 0;
                console.log(utils_1.LOG.VERSION_CREATED(versionNumber));
                _b.label = 5;
            case 5:
                utils_1.spinner.setSpinnerTitle(utils_1.LOG.DEPLOYMENT_CREATE);
                if (!!deploymentId) return [3 /*break*/, 7];
                return [4 /*yield*/, auth_1.script.projects.deployments.create({
                        scriptId: scriptId,
                        requestBody: {
                            versionNumber: versionNumber,
                            manifestFileName: utils_1.PROJECT_MANIFEST_BASENAME,
                            description: description,
                        },
                    })];
            case 6:
                // if no deploymentId, create a new deployment
                deployments = _b.sent();
                return [3 /*break*/, 9];
            case 7: return [4 /*yield*/, auth_1.script.projects.deployments.update({
                    scriptId: scriptId,
                    deploymentId: deploymentId,
                    requestBody: {
                        deploymentConfig: {
                            versionNumber: versionNumber,
                            manifestFileName: utils_1.PROJECT_MANIFEST_BASENAME,
                            description: description,
                        },
                    },
                })];
            case 8:
                // elseif, update deployment
                deployments = _b.sent();
                _b.label = 9;
            case 9:
                utils_1.spinner.stop(true);
                if (deployments.status !== 200)
                    utils_1.logError(null, utils_1.ERROR.DEPLOYMENT_COUNT);
                console.log("- " + deployments.data.deploymentId + " @" + versionNumber + ".");
                return [2 /*return*/];
        }
    });
}); });
