// @google/clasp/src/inquirer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var inquirer_1 = require("inquirer");
var apis_1 = require("./apis");
var utils_1 = require("./utils");
inquirer_1.registerPrompt('autocomplete', require('inquirer-autocomplete-prompt'));
/**
 * Inquirer prompt for a functionName.
 * @returns {Promise<{ functionName: string }>} A promise for an object with the `functionName` property.
 */
exports.functionNamePrompt = function (source) {
    var question = {
        name: 'functionName',
        message: 'Select a functionName',
        type: 'autocomplete',
        source: source,
    };
    return inquirer_1.prompt([question]);
};
/**
 * Inquirer prompt for a deployment Id.
 * @param {DeploymentIdPrompt[]} choices An array of `DeploymentIdPrompt` objects.
 * @returns {Promise<{ deploymentId: string }>} A promise for an object with the `deploymentId` property.
 */
exports.deploymentIdPrompt = function (choices) {
    return inquirer_1.prompt([{
            choices: choices,
            message: 'Open which deployment?',
            name: 'deployment',
            type: 'list',
        }]);
};
/**
 * Inquirer prompt for a project description.
 * @returns {Promise<{ description: string }>} A promise for an object with the `description` property.
 */
exports.descriptionPrompt = function () { return inquirer_1.prompt([{
        default: '',
        message: utils_1.LOG.GIVE_DESCRIPTION,
        name: 'description',
        type: 'input',
    }]); };
/**
 * Inquirer prompt for oauth scopes.
 * @returns {Promise<PromptAnswers>} A promise for an object with the `PromptAnswers` interface.
 */
exports.oauthScopesPrompt = function () { return inquirer_1.prompt([{
        message: 'Authorize new scopes?',
        name: 'doAuth',
        type: 'confirm',
    }, {
        message: 'Use localhost?',
        name: 'localhost',
        type: 'confirm',
        when: function (answers) {
            return answers.doAuth;
        },
    }]); };
/**
 * Inquirer prompt for overwriting a manifest.
 * @returns {Promise<{ overwrite: boolean }>} A promise for an object with the `overwrite` property.
 */
exports.overwritePrompt = function () { return inquirer_1.prompt([{
        default: false,
        message: 'Manifest file has been updated. Do you want to push and overwrite?',
        name: 'overwrite',
        type: 'confirm',
    }]); };
/**
 * Inquirer prompt for project Id.
 * @returns {Promise<{ projectId: string }>} A promise for an object with the `projectId` property.
 */
exports.projectIdPrompt = function () { return inquirer_1.prompt([{
        message: "" + utils_1.LOG.ASK_PROJECT_ID,
        name: 'projectId',
        type: 'input',
    }]); };
/**
 * Inquirer prompt for script Id.
 * @param {ScriptIdPrompt[]} fileIds An array of `ScriptIdPrompt` objects.
 * @returns {Promise<{scriptId: string;}>} A promise for an object with the `scriptId` property.
 */
exports.scriptIdPrompt = function (fileIds) { return inquirer_1.prompt([
    {
        choices: fileIds,
        message: utils_1.LOG.CLONE_SCRIPT_QUESTION,
        name: 'scriptId',
        pageSize: 30,
        type: 'list',
    },
]); };
/**
 * Inquirer prompt for script type.
 * @returns {Promise<{ type: string }>} A promise for an object with the `type` property.
 */
exports.scriptTypePrompt = function () { return inquirer_1.prompt([{
        choices: Object.keys(apis_1.SCRIPT_TYPES).map(function (key) { return apis_1.SCRIPT_TYPES[key]; }),
        message: utils_1.LOG.CREATE_SCRIPT_QUESTION,
        name: 'type',
        type: 'list',
    }]); };
