// @google/clasp/src/urls.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var apis_1 = require("./apis");
/**
 * Extracts scriptId from URL if given in URL form.
 * @param scriptId {string} either a scriptId or URL containing the scriptId
 * @example
 * extractScriptId(
 * 'https://script.google.com/a/DOMAIN/d/1Ng7bNZ1K95wNi2H7IUwZzM68FL6ffxQhyc_ByV42zpS6qAFX8pFsWu2I/edit'
 * )
 * returns '1Ng7bNZ1K95wNi2H7IUwZzM68FL6ffxQhyc_ByV42zpS6qAFX8pFsWu2I'
 * @example
 * extractScriptId('1Ng7bNZ1K95wNi2H7IUwZzM68FL6ffxQhyc_ByV42zpS6qAFX8pFsWu2I')
 * returns '1Ng7bNZ1K95wNi2H7IUwZzM68FL6ffxQhyc_ByV42zpS6qAFX8pFsWu2I'
 */
exports.extractScriptId = function (scriptId) {
    if (scriptId.length !== apis_1.SCRIPT_ID_LENGTH) {
        var ids = scriptId.split('/').filter(function (s) {
            return s.length === apis_1.SCRIPT_ID_LENGTH;
        });
        if (ids.length) {
            scriptId = ids[0];
        }
    }
    return scriptId;
};
// Helpers to get Apps Script project URLs
exports.URL = {
    APIS: function (projectId) { return "https://console.developers.google.com/apis/dashboard?project=" + projectId; },
    CREDS: function (projectId) { return "https://console.developers.google.com/apis/credentials?project=" + projectId; },
    LOGS: function (projectId) {
        return "https://console.cloud.google.com/logs/viewer?project=" + projectId + "&resource=app_script_function";
    },
    SCRIPT_API_USER: 'https://script.google.com/home/usersettings',
    // It is too expensive to get the script URL from the Drive API. (Async/not offline)
    SCRIPT: function (scriptId) { return "https://script.google.com/d/" + scriptId + "/edit"; },
    DRIVE: function (driveId) { return "https://drive.google.com/open?id=" + driveId; },
};
